<!DOCTYPE html>
<html>
  <head>
    <title>Promise</title>
  </head>
  <body>
    <button onclick="getPosts()">Get Posts</button>
    <button onclick="sendPost()">Send Posts</button>
  </body>
</html>
<script>
  /////////////////////////////////////////////

  // using promises and then
  const handleSuccess = async (response) => {
    console.log(
      `Request ${response.url} succesfully executed:`,
      await response.json()
    );
  };
  const handleServerError = (response) => {
    throw new Error(response.status);
  };

  const isSuccessfulResponse = (response) => {
    return response.ok;
  };

  const getPosts = () => {
    let url = "https://jsonplaceholder.typicode.com/postsass";
    // url = "https://jsonplaceholder.typicode.com/sfasf"; // server error would be thrown
    // url = "asd"; // default error in catch block would be thrown
    fetch(url)
      .then((response) => {
        isSuccessfulResponse(response)
          ? handleSuccess(response)
          : handleServerError(response);
      })
      .catch((error) => {
        // an error is already thrown by default. Implement a specific function if you need to handle it
      });
  };

  // using promises and async/await ()
  const sendPost = async (title, message) => {
    const body = {
      title,
      body: message,
      userId: 1,
    };

    try {
      const response = await fetch(
        "https://jsonplaceholder.typicode.com/posts",
        {
          method: "POST",
          body: JSON.stringify(body),
        }
      );

      isSuccessfulResponse(response)
        ? handleSuccess(response)
        : handleServerError(response);
    } catch (error) {
      // an error is already thrown by default. Implement a specific function if you need to handle it
    }
  };
</script>
